### PR TITLE
OCL VPP mosaic/flip/rotate implementation

### PR DESCRIPTION
**Pull request by dspmeng**
1. ocl: flip and rotate transform
    * support flip horizontal and vertical and rotate clockwise of 90, 180 and 270 degrees.
2. ocl: implement mosaic filter
    * apply mosaic filter using local average of blocks.
    * API update to 0.1.3.
3. example: enable mosaic and transform 
    * enable mosaic and transform to example blend, and add command arguments to select different tests
4. ocl: install ocl kernels to datadir

Test command:
$ ./blend -o --ot 128 -m --ms 32 --flip 1 --rotate 270 some.h264

### CODE REVIEW

File: vpp/oclpostprocess_transform.cpp [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        default:
+            m_transform = VPP_TRANSFORM_ROT_180;
+            break;
+        }
+    }
+
+    bool bFlip = m_transform & (VPP_TRANSFORM_FLIP_H | VPP_TRANSFORM_FLIP_V);
+    bool bRotate = m_transform & (VPP_TRANSFORM_ROT_90 | VPP_TRANSFORM_ROT_180 | VPP_TRANSFORM_ROT_270);
+
+    if (bFlip && bRotate) {
```
Comment (xuguangxin) [inline vpp/oclpostprocess_transform.cpp]: We did not optimize this combine. Why we need this together. Could user call flip then rotate?

File: vpp/oclpostprocess_transform.cpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+YamiStatus
+OclPostProcessTransform::process(const SharedPtr<VideoFrame>& src,
+    const SharedPtr<VideoFrame>& dst)
+{
+    YamiStatus status = ensureContext("transform");
+    if (status != YAMI_SUCCESS)
+        return status;
+
+    if (src->fourcc != YAMI_FOURCC_NV12 || dst->fourcc != YAMI_FOURCC_NV12) {
+        ERROR("only support transform of NV12 video frame");
```
Comment (xuguangxin) [inline vpp/oclpostprocess_transform.cpp]: could we add some width height check here? 320x240 can't rotate 320x240.. 320x240 can't flip to 120x120...

File: examples/blend.cpp [hunk 3]
```diff
@@ -80,12 +81,82 @@ class Blend
...
+                usage();
+                return false;
+            }
+        }
+        printf("Blend: %d, OSD(%d): %d, Mosaic(%d): %d, Transform: %u\n",
+            m_bBlend, m_osdThreshold, m_bOsd, m_mosaicSize, m_bMosaic, m_flipRot);
+
         m_input.reset(new VppInputDecode);
-        if (!m_input->init(argv[1])) {
+        if (!m_input->init(argv[argc - 1])) {
```
Comment (xuguangxin) [inline examples/blend.cpp]: better not assume last option is file name. you can use optind as https://github.com/xuguangxin/libyami/blob/master/examples/grid.cpp#L707

File: examples/blend.cpp [hunk 4]
```diff
@@ -119,33 +206,70 @@ class Blend
...
+                    SharedPtr<VideoFrame>& src = m_blendSurfaces[i];
+                    m_blender->process(src, m_dest);
+                }
+                PERF_STOP(blend);
+            }
+
+            if (m_bOsd) {
+                PERF_START(osd);
+                //osd
+                for (uint32_t i = 0; i < m_osdSurfaces.size(); i++) {
```
Comment (xuguangxin) [inline examples/blend.cpp]: duplicate code. could we add one function like
DrawSrufaces(m_osdSurfaces, m_osdBumpBoxes, m_osd)

File: examples/blend.cpp [hunk 5]
```diff
@@ -119,33 +206,70 @@ class Blend
...
+            if (m_bMosaic) {
+                PERF_START(mosaic);
+                //mosaic
+                for (uint32_t i = 0; i < m_mosaicBumpBoxes.size(); i++) {
+                    m_mosaicBumpBoxes[i]->getPos(m_dest->crop.x, m_dest->crop.y, m_dest->crop.width, m_dest->crop.height);
 
-                SharedPtr<VideoFrame>& src = m_osdSurfaces[i];
-                m_osd->process(src, m_dest);
+                    SharedPtr<VideoFrame> empty;
+                    m_mosaic->process(empty, m_dest);
```
Comment (xuguangxin) [inline examples/blend.cpp]: should call it like this m_mosaic->process(m_dest, m_dest);.So you did not create empty surface. for mosic operation m_dest is src and dest.

File: vpp/oclpostprocess_mosaic.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+#include "vaapipostprocess_factory.h"
+#include "common/common_def.h"
+#include "common/log.h"
+#include "ocl/oclcontext.h"
+#include "vpp/oclvppimage.h"
+
+namespace YamiMediaCodec {
+
+YamiStatus
+OclPostProcessMosaic::process(const SharedPtr<VideoFrame>& src,
```
Comment (xuguangxin) [inline vpp/oclpostprocess_mosaic.cpp]: better check src == dest.

File: vpp/oclpostprocess_transform.cpp (hunk 1 above)
Comment (dspmeng) [inline vpp/oclpostprocess_transform.cpp]: I intended to make the api flexible, so user cat toggle single or multiple flags in setParameters(). After your suggestion, I realize that the combined case can be optimized by adding 2 more ocl kernels, and also scratch buffer is unnecessary.

There are totally 7 cases.
case 1 (new kernel): FH+R90, FV+R270, R90+FV, R270+FH
case 2 (new kernel): FV+R90, FH+R270, R90+FH, R270+FV
case 3: FV, FH+R180, R180+FH
case 4: FH, FV+R180, R180+FV
case 5: R90
case 6: R180, FH+FV, FV+FH 
case 7: R270

I also assume Flip is performed before Rotate if user set both flags. So need to add a new private function flipAndRotate().

File: vpp/oclpostprocess_transform.cpp (hunk 2 above)
Comment (dspmeng) [inline vpp/oclpostprocess_transform.cpp]: I'll add width height check in private functions: flip, rotate, fliphAndRotate.

File: examples/blend.cpp (hunk 3 above)
Comment (dspmeng) [inline examples/blend.cpp]: done

File: examples/blend.cpp (hunk 4 above)
Comment (dspmeng) [inline examples/blend.cpp]: done

File: examples/blend.cpp (hunk 5 above)
Comment (dspmeng) [inline examples/blend.cpp]: done

File: vpp/oclpostprocess_mosaic.cpp (hunk 6 above)
Comment (dspmeng) [inline vpp/oclpostprocess_mosaic.cpp]: done

**Comment by dspmeng** [comment]
update patches according to review comments.

**Comment by dspmeng** [comment]
Finish update. Please review.

File: vpp/oclpostprocess_transform.cpp [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    }
+    else if (m_transform & VPP_TRANSFORM_ROT_180) {
+        if (width != dst->getWidth() || height != dst->getHeight()) {
+            ERROR("rotate failed due to unmatched resolution");
+            return YAMI_INVALID_PARAM;
+        }
+        size = 2;
+        w = width / 4 - 1;
+        h = height;
+        kernel = getKernel("transform_rot_180");
```
Comment (xuguangxin) [inline vpp/oclpostprocess_transform.cpp]: This maybe have performance issue, if you compare the string for every frame.best way is cache kernels. But you can update this later

**Comment by xuguangxin** [comment]
+1

File: vpp/oclpostprocess_transform.cpp (hunk 7 above)
Comment (dspmeng) [inline vpp/oclpostprocess_transform.cpp]: What about adding some private members to OclPostProcessTransform like:
SharedPtr<cl_kernel> m_kernelRot90;
SharedPtr<cl_kernel> m_kernelRot180;

And add a virtual function createKernels() to be called by OclPostProcessBase::ensureContext(). createKernels() will initialized each cl kernel private member like:
m_kernelRot90.reset(new cl_kernel(getKernel("transform_rot_90")), OclKernelDeleter);

If this is OK, I'll update in next pull request.

Comment (xuguangxin) [inline vpp/oclpostprocess_transform.cpp]: seems you do not need  hold the cl_kernel reference. The OclPostProcessBase will help you load and release kernels. Add pure virtual function prepareKernels in OclPostProcessBase ,and call it in in ensureContext  maybe better.

**Comment by xuguangxin** [comment]
merged as f8de055a25ee79bf71e5e54e871b6c4ce3a2349c.

**Pull request closed without merging**